### PR TITLE
Support pathfinding avoiding sharp objects (barbed wire etc)

### LINKED ISSUE #37928: Auto-planting gratuitously paths into dangerous terrain
# Describe the bug

When using (O) to Far(m) plots, the pathing for the player can be taken into dangerous territory when completely unnecessary. This can generate the "Really step into barbed wire fence?" type of message a LOT, especially if the farm plot is bordered by dangerous territory like a barbed wire fence.

# Steps To Reproduce

Two basic use cases that feel like they should be dealt with:

- Create a farm plot tightly bordered by barbed wire (e.g. a 3x3 plot, bordered by a 5x5 barbed wire fence).
- Zone it as a plot, then (O) and (m) to "Farm plots" while inside the plot and holding the given seeds.
- Very often, you will get interrupted with the "Really step into barbed wire fence?" message when attempting to plant the barbed-wire-bordered tiles because the algorithm is trying to place you on the fence tiles to plant.

As a real-world example, in this screenshot, the player has two isolated squares to the west that are plantable. They are zoned as farm plot. In many cases, doing O-m will immediately generate the "Really...?" message. (Probability seems roughly equal to the distribution of adjacent tiles that are fence -- e.g. 3/8 for edge-facing squares, 5/8 for corner squares)

image

OR

- Create a farm plot with a barbed wire fence on one border (e.g. a 3x3 plot, with a 3-long barbed wire fence to the direct east).
- Zone it as a plot for a given seed. Stand directly east of the fence with the seeds in hand.
- Do (O) and (m) to "Farm plots".
- The character will not do a simple path around the fence, you will get the "Really step into barbed wire fence?" message.

Same happens if the needed tool (a digging tool, if the earth is untilled) is on the other side of the fence.

Another real-world screenshot of the above: in this screenshot, the digging tool is to my east-southeast outside of the fence, the plot is to my west. There is a safe gap in the fence, but (m) does not path through that, it gives the "Really...?" message 100% of the time.

image

# Expected behavior

In case 1 above, it seems pretty simple that if there is a destination planting square, the game should not attempt to place me on the most dangerous adjacent tile to plant that square. It should select the safest adjacent tile to do the planting. This would prevent the "Really...?" message when the border-level tiles get planted in most cases.

Case 2 is obviously a LOT more complex, and less critical, but ideally I should select a path that is safe if it's not massively out of the way of the most direct route. This will take some cost-based pathing but this same issue has come up in other issues like auto-move so maybe will be re-usable logic?

# Screenshots

See above.

# Versions and configuration

- OS: Windows
    - OS Version: 10.0 1903
- Game Version: 0.D-11809-g78e6e40 [64-bit]
- Graphics Version: Tiles
- Mods loaded: [
    Dark Days Ahead [dda],
    Disable NPC Needs [no_npc_food],
    Beta National Guard Camp [national_guard_camp],
    More Locations [more_locations],
    Fuji's More Buildings [FujiStruct],
    More City Locations [cityside],
    Urban Development [Urban_Development],
    No Fungal Monsters [No_Fungi]
]

### PR DESCRIPTION
**Pull request by jkraybill**
#### Summary
```SUMMARY: Bugfixes "Avoid sharp things in player pathfinding"```

#### Purpose of change
Fixes #37928 by preventing the player from auto-pathing into sharp things (barbed wire fences etc).

#### Describe the solution
I added a new pathfinding setting, to avoid sharp things, and set it to default true for the player and NPCs and false for monsters.

I made sharp items unpathable if avoid_sharp=true, because there is an issue with route_adjacent in activity_item_handling.cpp. In farming, route_adjacent is called. That method looks up the closest adjacent square to the target "to be planted" square, and if there is a valid path, regardless of cost, it will go to the first valid adjacent square. So I couldn't just give sharp objects a high cost, because path cost isn't taken into account in route_adjacent if the high-cost tile is the destination adjacent tile.

#### Describe alternatives you've considered
I considered just adding sharp-object-avoidance to the avoid_traps behaviour. This still seems like a reasonable alternative if there is going to never be a situation where pathfinding should avoid either traps or sharp things but not the other.

#### Testing
I did quite a lot of testing with pathfinding related to farming automation. I did not do a lot of testing related to non-farm-automation pathfinding.

Example of the types of setups I was testing:

image

#### Additional context
This is my second C++ commit of any significance, I've picked a couple doozies between fuel consumption and pathfinding. I would not be at all surprised if I have made some rookie mistakes and welcome all feedback.

I also expect, but have not confirmed, that this commit will fix some other annoyances with player auto-pathfinding into bushes and barbed wires.

### CODE REVIEW

**Comment by jkraybill** [comment]
Moving this to WIP because there's now some weird unintended behaviour in manual movement. Working on it.

**Comment by jkraybill** [comment]
Removed WIP tag, the test cases I was using are replicable in master with pit traps and auto-travel.

**Comment by jkraybill** [comment]
Did significantly more testing with overmap auto-move. This PR appears to fully resolve the "raspberry bush problem" at the expense of making raspberry bushes and other sharp objects effectively impassable terrain for automove. I think that's a good thing but it's entirely possible I have missed an edge case. This PR makes territory fully bordered by sharp objects un-navigable by auto-move, so if that causes problems I'm not aware of, please let me know.

**Comment by davidpwbrown** [comment]
> Did significantly more testing with overmap auto-move. This PR appears to fully resolve the "raspberry bush problem" at the expense of making raspberry bushes and other sharp objects effectively impassable terrain for automove. I think that's a good thing but it's entirely possible I have missed an edge case. This PR makes territory fully bordered by sharp objects un-navigable by auto-move, so if that causes problems I'm not aware of, please let me know.

I think thats fine, automove shouldnt do anything risky or annoying, the player can do that themselves if they want to.

**Comment by jeremyshannon** [comment]
>false for monsters.
I feel like that should only be false for unintelligent monsters like zombies or the blob. It's hard to have a pet dog right now when they routinely impale themselves on barbed wire and jagged wreckage because they won't path around it.

**Comment by jkraybill** [comment]
@jeremyshannon thanks for the tip. Opening a separate issue for that, this PR does add a JSON flag to allow that to be specified per monster.

File: src/pathfinding.h [hunk 1]
```diff
@@ -60,12 +61,15 @@ struct pathfinding_settings {
     bool avoid_traps = false;
     bool allow_climb_stairs = true;
     bool avoid_rough_terrain = false;
+    bool avoid_sharp = false;
 
     pathfinding_settings() = default;
     pathfinding_settings( const pathfinding_settings & ) = default;
-    pathfinding_settings( int bs, int md, int ml, int cc, bool aod, bool at, bool acs, bool art )
+    pathfinding_settings( int bs, int md, int ml, int cc, bool aod, bool at, bool acs, bool art,
+                          bool as )
```
Comment (codemime) [inline src/pathfinding.h]: The fifth boolean argument is a little bit too much here (actually, even a single one is usually evil). It's impossible to figure the meaning of the train of `true`s and `false`s without looking at the declaration. I'd suggest either dropping all the constructors in favor of explicit assignments:

```c++
pathfinding_settings ps;
ps.allow_open_doors = true;
ps.avoid_traps = false;
...
```
or using a bitset of named flags for this purpose.

Comment (jkraybill) [reply src/pathfinding.h]: thanks, I'll open an issue for this, this probably could be done for a number of methods like this.

**Pull request merged**